src/omp/backend.f90

```diff
@@ -168,61 +170,69 @@ subroutine transeq_omp_dist(self, du, dv, dw, u, v, w, dirps)
     class(field_t), intent(in) :: u, v, w
     type(dirps_t), intent(in) :: dirps
 
-    call transeq_halo_exchange(self, u, v, w, dirps)
+
+
+    call transeq_halo_exchange(self, u, v, w, dirps%dir)
 
     call transeq_dist_component(self, du, u, u, &
                                 self%u_recv_s, self%u_recv_e, &
                                 self%u_recv_s, self%u_recv_e, &
                                 dirps%der1st, dirps%der1st_sym, &
-                                dirps%der2nd, dirps)
+                                dirps%der2nd, dirps%dir)
     call transeq_dist_component(self, dv, v, u, &
                                 self%v_recv_s, self%v_recv_e, &
                                 self%u_recv_s, self%u_recv_e, &
                                 dirps%der1st_sym, dirps%der1st, &
-                                dirps%der2nd_sym, dirps)
+                                dirps%der2nd_sym, dirps%dir)
     call transeq_dist_component(self, dw, w, u, &
                                 self%w_recv_s, self%w_recv_e, &
                                 self%u_recv_s, self%u_recv_e, &
                                 dirps%der1st_sym, dirps%der1st, &
-                                dirps%der2nd_sym, dirps)
+                                dirps%der2nd_sym, dirps%dir)
 
   end subroutine transeq_omp_dist
 
-  subroutine transeq_halo_exchange(self, u, v, w, dirps)
+  subroutine transeq_halo_exchange(self, u, v, w, dir)
     class(omp_backend_t) :: self
     class(field_t), intent(in) :: u, v, w
-    type(dirps_t), intent(in) :: dirps
-    integer :: n_halo
+    integer, intent(in) :: dir
+    integer :: n_halo, n, nproc_dir, pprev, pnext
+    integer :: n_groups
 
     ! TODO: don't hardcode n_halo
     n_halo = 4
+    n_groups = self%mesh%get_n_groups(dir)
+    n = self%mesh%get_n(u)
```

Comment:
actually, I am tempted to keep it here. This is a halo swap which has nothing to do with the tri diagonal system. So I think it is better to keep `get_n`

Reply:
I see, in `tds_solve` we have the operator so can easily use the `tds_n` but here we don't have an operator available.

The only issue is `get_n` from a field requires the field to have a defined `data_loc`. in `DIR_X` this is fine because the `data_loc` for `solver%u/v/w` are all defined in this PR. But we call this function in `DIR_Y` and `DIR_Z` as well, and then it will complain about a `NONE` `data_loc` and won't be able to get the `n`.

What about using `self%mesh%get_n(dir, VERT)` here instead? transeq_component is a specialist implementation and we know that it always works from `VERT` to `VERT`, so it will solve the problem. Not the best solution, but maybe we can try something else later in #98, what do you think?

Reply:
Just defined `data_loc` in a lot more fields in `solver.f90` so it should be fine. In preparation of #98, I only defined it for input fields so it shouldn't be redundant.

Reply:
Yes, this solves the problem.

Just to clarify, with #98 we'll be able to obtain `data_loc`s in all the fields in solver by specifying `data_loc` of the `solver%u`, `solver%v`, and `solver%w` only. Because in the rest we'll be able to obtain the resulting `data_loc` from the operations, either from `tds_solve` or `reorder`. For example for `u_y` in `transeq`, we `reorder` data from `u` to `u_y`, so `reorder` function can copy the `data_loc` as well as reordering and copying the data inside. So specifying the `data_loc` for `u_y` and many others will be redundant with #98.

Reply:
ah yes indeed. I will let you remove the redundant ones then if you feel that's needed.